multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +944,365 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+	// object -- for headless services, it takes a while for the backing pod
+	// set to come up (especially in the case of StatefulSets). Because of
+	// this, we may not have any addresses in the Endpoints when we export
+	// (and create) the headless service mirror; instead, we leave the
+	// Endpoints creation to the Endpoints informer. If we receive an
+	// Endpoints update on the remote cluster, and the mirrorerd version of
+	// the Endpoints does not exist then it means we need to create it (and
+	// all of its nested clusterIP services).
+	if err != nil {
+		// If the error we get when we try to list the local Endpoints object
+		// (for the headless service) is not a NotFound error, then it means
+		// something _actually_ went wrong, so return.
+		if !kerrors.IsNotFound(err) {
+			return err
+		}
+
+		// If the error is NotFound, it means the remote Endpoints object has
+		// finally been updated to include at least one address so we can start
+		// building the Endpoints object for the mirror. We need to get the
+		// original service first so we can create annotations for the Endpoints
+		// object and all other clusterIP services (one clusterIP per address in
+		// Endpoints).
+		remoteService, err := rcsw.remoteAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(remoteEp.Name)
+		if err != nil {
+			return err
+		}
+		localServiceName := rcsw.mirroredResourceName(remoteService.Name)
+		serviceInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, remoteService.Name)
+
+		// To mirror a headless service, we need to create a clusterIP for each
+		// host in its Endpoints.
+		// epSet is a map that holds all of the Endpoints hosts to be created in
+		// the src cluster, with keys set to the clusterIP serviceName and the
+		// value to its clusterIP in the src cluster.
+		epSet := make(map[string]string)
+		// For each host, we want to create a service. All services will be set
+		// as hosts in the headless service's Endpoints object (in the src
+		// cluster). For prototype, assume only one subset.
+		for _, addr := range remoteEp.Subsets[0].Addresses {
+
+			// Create annotations for host's service
+			ref := addr.TargetRef
+			if ref == nil {
+				// ignore addresses w/o a backing pod (if they exist)
+				continue
+			}
+			// Name of the backing targetRef, generally a *pod*
+			refLocalName := rcsw.mirroredResourceName(ref.Name)
+			err := rcsw.createHeadlessMirroredHost(ctx, ref, remoteService, refLocalName)
+			if err != nil {
+				rcsw.log.Errorf("failed to create headless mirrored hosts for %s: %v", refLocalName, err)
+			}
+
+			createdService, err := rcsw.localAPIClient.Client.CoreV1().Services(remoteService.Namespace).Get(ctx, refLocalName, metav1.GetOptions{})
+			if err != nil {
+				rcsw.log.Debugf("failed to list clusterIP Service for mirrored host %s/%s: %v", remoteEp.Namespace, addr.Hostname, err)
+				return err
+			}
+			epSet[ref.Name] = createdService.Spec.ClusterIP
```

Comment:
Once the service creation call returns, is it guaranteed a ClusterIP will be assigned already?

Reply:
Good question. I'm not sure but whenever I looked through my changes I always asked myself the same: "can we reliably get an IP address here without nil dereferences?". So far I haven't been able to break this.

I assume it wouldn't be a problem because clusterIP fields are immutable, so when a service is created we should have a clusterIP in its spec before we can see it. No way for me to test this assumption though and also didn't find anything in the docs.

Another issue I have here is the creation time of a service, does this have the potential for a race condition or falling out of sync with the API Server? Can we create a service and by the time we list it it hasn't been created yet so we get an error?